api/endpoints.go

```diff
@@ -825,3 +827,103 @@ func handleDailyRankingPageCount(w http.ResponseWriter, r *http.Request) {
 
 	_, _ = w.Write([]byte(strconv.Itoa(count)))
 }
+
+// redirect link after authorizing application link
+func handleProviderCallback(w http.ResponseWriter, r *http.Request) {
+	provider := r.PathValue("provider")
+	state := r.URL.Query().Get("state")
+	gameUrl := os.Getenv("GAME_URL")
+	var externalAuthId string
+	var err error
+	switch provider {
+	case "discord":
+		externalAuthId, err = account.HandleDiscordCallback(w, r)
+	case "google":
+		externalAuthId, err = account.HandleGoogleCallback(w, r)
+	default:
+		http.Error(w, "invalid provider", http.StatusBadRequest)
+		return
+	}
+
+	if err != nil {
+		httpError(w, r, err, http.StatusInternalServerError)
+		return
+	}
+
+	if state != "" {
+		state = strings.Replace(state, " ", "+", -1)
+		stateByte, err := base64.StdEncoding.DecodeString(state)
+		if err != nil {
+			defer http.Redirect(w, r, gameUrl, http.StatusSeeOther)
+			return
+		}
+
+		userName, err := db.FetchUsernameBySessionToken(stateByte)
+		if err != nil {
+			defer http.Redirect(w, r, gameUrl, http.StatusSeeOther)
+			return
+		}
+
+		switch provider {
+		case "discord":
+			err = db.AddDiscordIdByUsername(externalAuthId, userName)
+		case "google":
+			err = db.AddGoogleIdByUsername(externalAuthId, userName)
+		}
+
+		if err != nil {
+			defer http.Redirect(w, r, gameUrl, http.StatusSeeOther)
+			return
+		}
+
+	} else {
+		var userName string
+		switch provider {
+		case "discord":
+			userName, err = db.FetchUsernameByDiscordId(externalAuthId)
+		case "google":
+			userName, err = db.FetchUsernameByGoogleId(externalAuthId)
+		}
+		if err != nil {
+			defer http.Redirect(w, r, gameUrl, http.StatusSeeOther)
+			return
+		}
+
+		sessionToken, err := account.GenerateTokenForUsername(userName)
+		if err != nil {
+			defer http.Redirect(w, r, gameUrl, http.StatusSeeOther)
+			return
+		}
+
+		http.SetCookie(w, &http.Cookie{
+			Name:  "pokerogue_sessionId",
+			Value: sessionToken,
+			Path:  "/",
+		})
+	}
+
+	defer http.Redirect(w, r, gameUrl, http.StatusSeeOther)
```

Comment:
There is no defer needed here. THe function only has one return path from here and that is right in the next line